rust/src/librarydb.rs

```diff
@@ -0,0 +1,504 @@
+    base_path: String,
+    /// List of ok entries in the library.
+    entries: Vec<LibraryEntry>,
+}
+
+/// Library entry.
+/// Represents a file in a library.
+#[derive(Debug)]
+pub struct LibraryEntry {
+    /// File path without the base path of the library.
+    file_path: String,
+    /// Start of the file data in the library file.
+    data_start: u64,
+    /// End of the file data in the library file.
+    data_end: u64,
+}
+
+/// Library file.
+/// Provides access to the data of a library entry through Read and Seek.
+#[derive(Debug)]
+pub struct LibraryFile {
+    /// The index of the entry of this file.
+    index: usize,
+    /// The thread safe library of this file.
+    arc_library: Arc<RwLock<Library>>,
+    /// Current position in the library file.
+    position: u64,
+}
+
+impl LibraryDB {
+    /// Constructor.
+    pub fn new() -> Self {
+        Self {
+            inner: Mutex::new(LibraryDBInner::new()),
+        }
+    }
+
+    /// Returns thread safe guard that grants direct access.
+    /// While the guard lives, no one else has access to the inner library database.
+    pub fn locked(&self) -> MutexGuard<LibraryDBInner> {
+        self.inner.lock().unwrap()
+    }
+
+    /// Opens and adds a library at the end of library database.
+    pub fn add_library(&mut self, data_dir: &Path, library: &Path) -> io::Result<()> {
+        self.locked().add_library(data_dir, library)
+    }
+
+    /// Opens a library file for reading.
+    /// The file must be dropped before the library database is dropped.
+    pub fn open_file(&self, path: &str) -> io::Result<LibraryFile> {
+        self.locked().open_file(path)
+    }
+}
+
+impl LibraryDBInner {
+    /// Constructor.
+    fn new() -> Self {
+        Self {
+            arc_libraries: Vec::new(),
+        }
+    }
+
+    /// Opens and adds a library at the end of library database.
+    pub fn add_library(&mut self, data_dir: &Path, library: &Path) -> io::Result<()> {
+        let library = Library::open_library(&data_dir, &library)?;
+        self.arc_libraries.push(Arc::new(RwLock::new(library)));
+        Ok(())
+    }
+
+    /// Opens a library file for reading.
+    /// The file must be dropped before the library database is dropped.
+    pub fn open_file(&self, path: &str) -> io::Result<LibraryFile> {
+        let path = case_insensitive_path(&path);
+        for arc_library in &self.arc_libraries {
+            let library = arc_library.read().unwrap();
+            if let Some(index) = library.find_with_case_insensitive(&path) {
+                return Ok(LibraryFile {
+                    index,
+                    arc_library: arc_library.to_owned(),
+                    position: library.entries[index].data_start,
+                });
+            }
+        }
+        Err(io::ErrorKind::NotFound.into())
+    }
+}
+
+impl Library {
+    /// Opens a library for reading.
+    fn open_library(data_dir: &Path, library: &Path) -> io::Result<Self> {
```

Comment:
I would just call this `open` as `library` is already the struct name.

Reply:
Done.